tokio/src/sync/semaphore_ll.rs

```diff
@@ -501,86 +596,149 @@ impl Permit {
     ///
     /// The permit begins in the "unacquired" state.
     pub(crate) fn new() -> Permit {
+        use PermitState::Acquired;
+
         Permit {
             waiter: None,
-            state: PermitState::Idle,
+            state: Acquired(0),
         }
     }
 
     /// Returns true if the permit has been acquired
     pub(crate) fn is_acquired(&self) -> bool {
-        self.state == PermitState::Acquired
+        match self.state {
+            PermitState::Acquired(num) if num > 0 => true,
+            _ => false,
+        }
     }
 
     /// Try to acquire the permit. If no permits are available, the current task
     /// is notified once a new permit becomes available.
     pub(crate) fn poll_acquire(
         &mut self,
         cx: &mut Context<'_>,
+        num_permits: u16,
         semaphore: &Semaphore,
     ) -> Poll<Result<(), AcquireError>> {
+        use std::cmp::Ordering::*;
+        use PermitState::*;
+
         match self.state {
-            PermitState::Idle => {}
-            PermitState::Waiting => {
+            Waiting(requested) => {
+                // There must be a waiter
                 let waiter = self.waiter.as_ref().unwrap();
 
-                if waiter.acquire(cx)? {
-                    self.state = PermitState::Acquired;
+                match requested.cmp(&num_permits) {
+                    Less => {
+                        let delta = num_permits - requested;
+
+                        // Request additional permits. If the waiter has been
+                        // dequeued, it must be re-queued.
+                        if !waiter.try_inc_permits_to_acquire(delta as usize) {
+                            let waiter = NonNull::from(&**waiter);
+
+                            // Ignore the result. The check for
+                            // `permits_to_acquire()` will converge the state as
+                            // needed
+                            let _ = semaphore.poll_acquire2(delta, || Some(waiter))?;
+                        }
+
+                        self.state = Waiting(num_permits);
+                    }
+                    Greater => {
+                        let delta = requested - num_permits;
+                        let to_release = waiter.try_dec_permits_to_acquire(delta as usize);
+
+                        semaphore.add_permits(to_release);
+                        self.state = Waiting(num_permits);
+                    }
+                    Equal => {}
+                }
+
+                if waiter.permits_to_acquire()? == 0 {
```

Comment:
Why does this happen before and after `register_by_ref`? To avoid the register in the first place. And then there is a race between registering the waker and getting the permits from another thread?

Reply:
Yes, pretty much. Updating `AtomicWaker` is not *super* cheap.